File: solutions/heuristic/greedy_earliest_available_adaptive.py
from sys import argv

# Set to true to add additional text explaining the output, false otherwise
printed = True

with open(f"../../input/{int(argv[1])}.txt", "r") as f:
    n = int(f.readline().strip())
    e = [0] * (n + 1)
    l = [0] * (n + 1)
    d = [0] * (n + 1)

    for i in range(1, n + 1):
        e[i], l[i], d[i] = map(int, f.readline().split())
    t = [list(map(int, f.readline().split())) for _ in range(n + 1)]
visited = [False] * (n + 1)
previous_point = 0
total_time = 0
previous_times = [0] * (n + 1)
print(f"Problem size             : {n}" if printed else n)

if printed:
    print("Solution                 :", end=" ")

for i in range(1, n + 1):
    best_deadline = float("inf")
    best_distance = float("inf")

    for j in range(1, n + 1):
        if (
            not visited[j]
            and previous_times[i - 1] + t[previous_point][j] <= l[j]
            and (
                l[j] < best_deadline
                or (l[j] == best_deadline and t[previous_point][j] < best_distance)
            )
        ):
            best_point = j
            best_deadline = l[j]
            best_distance = t[previous_point][j]
    print(best_point, end=" ")
    visited[best_point] = True
    total_time += t[previous_point][best_point]
    previous_times[i] = (
        max(previous_times[i - 1] + t[previous_point][best_point], e[best_point])
        + d[best_point]
    )
    previous_point = best_point

if printed:
    print(
        f"\nTotal time               : {total_time + t[previous_point][0]}\nTime finishing delivering:",
        *previous_times[1:],
    )

File: solutions/heuristic/greedy_earliest_available_adaptive.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

int main(int argc, char *argv[])
{
    // Set to true to add additional text explaining the output, false otherwise
    bool print = true;

    // Comment out the following line to read from the input file instead of standard input
    freopen(("../../input/" + to_string(stoi(argv[1])) + ".txt").c_str(), "r", stdin);

    int n, previous_point = 0, total_time = 0;
    cin >> n;
    vector<int> e(n + 1), l(n + 1), d(n + 1), previous_times(n + 1);

    for (int i = 1; i <= n; i++)
    {
        cin >> e[i] >> l[i] >> d[i];
    }
    vector<vector<int>> t(n + 1, vector<int>(n + 1));

    for (int i = 0; i <= n; i++)
    {
        for (int j = 0; j <= n; j++)
        {
            cin >> t[i][j];
        }
    }

    if (print)
    {
        cout << "Problem size             : ";
    }
    cout << n << endl;

    if (print)
    {
        cout << "Solution                 : ";
    }
    vector<bool> visited(n + 1, false);
    previous_times[0] = 0;

    for (int i = 1; i <= n; i++)
    {
        int best_point;

        for (int j = 1, best_deadline = INT_MAX, best_distance = INT_MAX; j <= n; j++)
        {
            if (!visited[j] && previous_times[i - 1] + t[previous_point][j] <= l[j] && (l[j] < best_deadline || (l[j] == best_deadline && t[previous_point][j] < best_distance)))
            {
                best_point = j;
                best_deadline = l[j];
                best_distance = t[previous_point][j];
            }
        }
        cout << best_point << " ";
        visited[best_point] = true;
        total_time += t[previous_point][best_point];
        previous_times[i] = max(previous_times[i - 1] + t[previous_point][best_point], e[best_point]) + d[best_point];
        previous_point = best_point;
    }

    if (print)
    {
        cout << "\nTotal time               : " << total_time + t[previous_point][0] << "\nTime finishing delivering:";

        for (int i = 1; i <= n; i++)
        {
            cout << " " << previous_times[i];
        }
    }
    cout << endl;
    return 0;
}